dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c

```diff
@@ -0,0 +1,289 @@
+    RelativePath.MaximumLength = RelativePath.Length;
+
+    /* If no volume was requested, return the relative path */
+    if (VolumeType == VOLUME_NAME_NONE)
+    {
+        FinalLength = RelativePath.Length / sizeof(WCHAR) + 1;
+        if (cchFilePath < FinalLength + 1)
+        {
+            SetLastError(ERROR_NOT_ENOUGH_MEMORY);
+            return 0;
+        }
+
+        /* Copy the name to the caller's buffer */
+        lpszFilePath[0] = L'\\';
+        RtlCopyMemory(&lpszFilePath[1], RelativePath.Buffer, RelativePath.Length);
+        lpszFilePath[FinalLength] = UNICODE_NULL;
+
+        /* Return the length of the name */
+        return FinalLength;
+    }
+
+    /* Query the DOS volume name */
+    RtlInitEmptyUnicodeString(&VolumeName, VolumeBuffer, sizeof(VolumeBuffer));
+    Status = QueryDosVolumeNameForNtDeviceName(&DeviceName, &VolumeName);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("QueryDosVolumeNameForNtDeviceName failed: %lx\n", Status);
+        BaseSetLastNTError(Status);
+        return 0;
+    }
+
+    if (VolumeType == VOLUME_NAME_DOS)
+    {
+        /* Calculate the final length and vaidate the buffer size */
+        FinalLength = ((VolumeName.Length + RelativePath.Length) / sizeof(WCHAR));
+        if (cchFilePath < FinalLength + 1)
+        {
+            SetLastError(ERROR_NOT_ENOUGH_MEMORY);
+            return FinalLength + 1;
+        }
+
+        /* Construct the final name */
+        RtlCopyMemory(lpszFilePath, VolumeName.Buffer, VolumeName.Length);
+        RtlCopyMemory(lpszFilePath + (VolumeName.Length / sizeof(WCHAR)),
+                      RelativePath.Buffer,
+                      RelativePath.Length);
+        lpszFilePath[FinalLength] = UNICODE_NULL;
+
+        /* Return the length of the name */
+        return FinalLength;
+    }
+    else if (VolumeType == VOLUME_NAME_GUID)
+    {
+        /* Query the GUID volume name */
+        Success = GetVolumeNameForVolumeMountPointW(VolumeBuffer,
+                                                    VolumeBuffer,
+                                                    ARRAYSIZE(VolumeBuffer));
+        if (!Success)
+        {
+            DPRINT1("GetVolumeNameForVolumeMountPointW failed: %d\n", GetLastError());
+            SetLastError(ERROR_PATH_NOT_FOUND);
+            return 0;
+        }
+
+        VolumeLength = wcslen(VolumeBuffer);
+        FinalLength = VolumeLength + (RelativePath.Length / sizeof(WCHAR));
+        if (cchFilePath < FinalLength + 1)
+        {
+            SetLastError(ERROR_NOT_ENOUGH_MEMORY);
+            return FinalLength + 1;
+        }
+
+        RtlCopyMemory(lpszFilePath, VolumeBuffer, VolumeLength * sizeof(WCHAR));
+        RtlCopyMemory(lpszFilePath + VolumeLength,
+                      RelativePath.Buffer,
+                      RelativePath.Length);
+        lpszFilePath[FinalLength] = UNICODE_NULL;
+
+        /* This is based on Windows behavior */
+        SetLastError(0);
```

Comment:
Since this is the only case where this appears to be done, I'd assume it comes from `GetVolumeNameForVolumeMountPointW`